### ISSUE TITLE
mixed grade elements

### ISSUE THREAD

**Issue by alok** [eschnett/DifferentialForms.jl#8]
can one create an element such as `2 + 3e1e2`?

**Comment by eschnett** [comment]
Yes, this is possible, although there isn't a nice syntax for it (yet).

`DifferentialForms` distinguishes between the types `Form` and `Multivector`. A form has a particular rank, whereas a multivector is the tensor sum of forms of all permitted ranks.

The implementation is optimized for small number of dimensions where type stability is important. The representation of a multivector encodes in its type (in a mask `M`) which elements can be non-zero, and only these are stored.

```Julia
julia> using DifferentialForms

julia> e = unit(Multivector{2,UInt64(0b0001),Float64})
Float64⟦1.0⟧{D=2,M=0b1}

julia> e1 = unit(Multivector{2,UInt64(0b0010),Float64}, 1)
Float64⟦1.0⟧{D=2,M=0b10}

julia> e2 = unit(Multivector{2,UInt64(0b0100),Float64}, 2)
Float64⟦1.0⟧{D=2,M=0b100}

julia> e1 ∧ e2
Float64⟦1.0⟧{D=2,M=0b1000}

julia> 2 * e + 3 * e1 ∧ e2
Float64⟦2.0,3.0⟧{D=2,M=0b1001}
```

I have used the `Form` type in production. I have not yet used `Multivector` in production, but there are extensive tests. I assume this means that there might be snags, and the syntax is not yet streamlined. For example, the way to create unit vectors above is still much too verbose.

**Comment by chakravala** [comment]
eschnett is an arrogant elitist prick of the corrupt academic type, just a warning @alok if you consider associating with him, he's the kind of abusive person who takes advantage of people and then backstab them for his own benefit

**Comment by alok** [comment]
what sort of syntax would you go for?

also, what do you think of using bitvectors for the mask?

**Comment by eschnett** [comment]
I'm open to ideas for syntax. For example, `unit(Multivector{2}, 1,2)` would already work for me, and that would be straightforward to implement.

I chose integers over bitvectors so that the mask is a bitstype. I don't know whether type parameter can be non-bitstype constants. I also assumed this would make things faster; I have not tested this, though. Do you have experience with this?
